### PR TITLE
Feature: QNX Neutrino support

### LINKED ISSUE #4294: QNX Neutrino support
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

there is some existing demand for QNX Neutrino support:
https://github.com/conan-community/community/issues/78
https://github.com/conan-io/conan/issues/2970

add an official support for QNX Neutrino operation system. it means, at least, the following:
- define new OS in `settings.yml` (name might be `QNX` or `Neutrino`)
- define OS versions, at least `6.5`, `6.6`, `7.0`
- define new compiler - `QCC`, and its versions (`4.2.2`, `5.4.0`)
- add standard libraries for QNX/QCC: `gpp` (GNU C++), `cpp` (Dinkum C+) and its variations (like Embedded Dinkum C++, Dinkum Abridged C++, etc)
- define GNU triplets for QNX cross-compiling (check `config.guess`)
- add required changes to the CMake helper (see https://cmake.org/cmake/help/v3.1/manual/cmake-toolchains.7.html#id9)

probably nice to have, but hard without access to the actual system - I doubt we will add it, as we mainly targeting cross-compiling from Windows/Linux/OSX to QNX, not running conan on QNX itself:
- add detection for QNX OS and compiler

/cc @jpovedano @Johnnyxy (as requestors)

### PR DESCRIPTION
**Pull request by SSE4**
closes #4294 

Changelog: Feature: QNX Neutrino support
Docs: https://github.com/conan-io/docs/pull/XXXX

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

/cc @jpovedano @Johnnyxy

### CODE REVIEW

File: conans/client/conf/__init__.py [hunk 1]
```diff
@@ -43,7 +43,9 @@
     SunOS:
     Arduino:
         board: ANY
-arch: [x86, x86_64, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr]
+    Neutrino:
+        version: ["6.4.0", "6.5.0", "7.0.0"]
```
Comment (memsharded) [inline conans/client/conf/__init__.py]: Are compiled binaries incompatible with different versions of the OS?

File: conans/client/conf/__init__.py [hunk 2]
```diff
@@ -13,12 +13,12 @@
...
 os_build: [Windows, WindowsStore, Linux, Macos, FreeBSD, SunOS]
-arch_build: [x86, x86_64, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr]
+arch_build: [x86, x86_64, ppc32be, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr, sh4]
 
 # Only for building cross compilation tools, 'os_target/arch_target' is the system for
 # which the tools generate code
-os_target: [Windows, Linux, Macos, Android, iOS, watchOS, tvOS, FreeBSD, SunOS, Arduino]
-arch_target: [x86, x86_64, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr]
+os_target: [Windows, Linux, Macos, Android, iOS, watchOS, tvOS, FreeBSD, SunOS, Arduino, Neutrino]
+arch_target: [x86, x86_64, ppc32be, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr, sh4]
```
Comment (memsharded) [inline conans/client/conf/__init__.py]: Are those architectures closely related to QNX Neutrino? Maybe could go to another PR?

File: conans/test/unittests/client/build/compiler_flags_test.py [hunk 3]
```diff
@@ -77,6 +78,21 @@ def test_libcxx_flags(self):
         arch_flags = libcxx_flag(compiler='sun-cc', libcxx='libstdc++')
         self.assertEquals(arch_flags, '-library=stdcpp')
 
+        arch_flags = libcxx_flag(compiler='sun-cc', libcxx='libstdc++')
```
Comment (memsharded) [inline conans/test/unittests/client/build/compiler_flags_test.py]: This seems duplicated

File: conans/client/conf/__init__.py (hunk 1 above)
Comment (SSE4) [reply conans/client/conf/__init__.py]: no idea, need to investigate it further

File: conans/test/unittests/client/build/compiler_flags_test.py (hunk 3 above)
Comment (SSE4) [reply conans/test/unittests/client/build/compiler_flags_test.py]: removed, thanks

File: conans/client/conf/__init__.py (hunk 2 above)
Comment (memsharded) [reply conans/client/conf/__init__.py]: - If ``ppcbe`` is implictly 32 bits, I am ok if the name doesn't contain 32.
- Why the arch is not called ``shle`` and called ``sh4``? Just asking, to understand.

Comment (SSE4) [reply conans/client/conf/__init__.py]: then I support better to name it `sh4le` (as there are SH1, SH2, SH3 and SH5, for example, so name just shle would be confusing)
we already have `ppc32` which is not just `ppc`, so probably it's more convenient to name it `ppc32be` then?

Comment (memsharded) [reply conans/client/conf/__init__.py]: Ok, good!

File: conans/client/conf/__init__.py (hunk 1 above)
Comment (SSE4) [reply conans/client/conf/__init__.py]: it seems to be forward-compatible (but not backward compatible)
http://www.qnx.com/developers/docs/6.5.0/index.jsp?topic=%2Fcom.qnx.doc.ide.userguide%2Ftopic%2Fmigrate_Binary_Compatibility.html

File: conans/client/conf/__init__.py (hunk 2 above)
Comment (SSE4) [reply conans/client/conf/__init__.py]: I've renamed sh4 -> sh4le

**Comment by Johnnyxy** [comment]
Hi there,
the setting for qcc: versions in settings.yml is not correct or something is missing.
QNX SDK 6.5 has the QCC compiler version 4.4.2 -> 4.4
QNX SDK 7.0 has the QCC compiler version 5.4.0 -> 5.4
Maybe older QNX SDKs have a 4.2 compiler version but I am missing 4.4.

:)

**Comment by SSE4** [comment]
@Johnnyxy thanks for the update, I'll change them

**Comment by memsharded** [comment]
There was an attempt to add QNX in https://github.com/conan-io/conan/pull/2971, by @jpovedano.

He was using versions 6.5 and 6.6, while you are using here 6.5.0 and 7.0.0, missing 6.6. Could you please check it? Also, if the patch is irrelevant, could we just drop it?

**Comment by SSE4** [comment]
@Johnnyxy @memsharded I've updated versions for QNX and QCC

**Comment by lasote** [comment]
We need more feedback from potential QNX users.

**Comment by Johnnyxy** [comment]
I will try to give detailed answers and hints for building for QNX Neutrino.

Prerequisites
-------------
Software for QNX Neutrino is being cross-compiled from Windows or Linux.

QNX Neutrino compiler
------------------------
The compiler for [QNX SDK 6.5](http://www.qnx.com/developers/docs/6.5.0/index.jsp?topic=%2Fcom.qnx.doc.neutrino_utilities%2Fq%2Fqcc.html) is defined as ```qcc``` for C and ```QCC``` for C++. In reality there is only one binary one can execute "qcc" (on Windows). I do not think that there are two binaries on Linux.
But on [QNX SDK 7.0](http://www.qnx.com/developers/docs/7.0.0/#com.qnx.doc.neutrino.utilities/topic/q/qcc.html) there are ```qcc``` for C and ```q++``` for C++ (no ```QCC``` anymore).

os_build and arch_build
------------------------
Typically one just needs os and arch for the desired QNX Neutrino target. Additionally os_build/os_arch are the conan supported ones. There is no need for os_build/arch_build or os_target/arch_target specifically for QNX Neutrino builds.

QNX SDK 6.5
-------------
The SDK supports the following target-runtime-combinations out-of-the-box (printed by command: ```qcc -V```):
- armle
- armle_acpp-ne
- armle_acpp
- armle_cpp-ne
- armle_cpp
- armle_ecpp-ne
- armle_ecpp
- armle_gpp
- armv7le
- armv7le_acpp-ne
- armv7le_acpp
- armv7le_cpp-ne
- armv7le_cpp
- armv7le_ecpp-ne
- armv7le_ecpp
- armv7le_gpp
- mipsbe
- mipsbe_acpp-ne
- mipsbe_acpp
- mipsbe_cpp-ne
- mipsbe_cpp
- mipsbe_ecpp-ne
- mipsbe_ecpp
- mipsbe_gpp
- mipsle
- mipsle_acpp-ne
- mipsle_acpp
- mipsle_cpp-ne
- mipsle_cpp
- mipsle_ecpp-ne
- mipsle_ecpp
- mipsle_gpp
- ppcbe
- ppcbespe
- ppcbespe_acpp-ne
- ppcbespe_acpp
- ppcbespe_cpp-ne
- ppcbespe_cpp
- ppcbespe_ecpp-ne
- ppcbespe_ecpp
- ppcbespe_gpp
- ppcbe_acpp-ne
- ppcbe_acpp
- ppcbe_cpp-ne
- ppcbe_cpp
- ppcbe_ecpp-ne
- ppcbe_ecpp
- ppcbe_gpp
- shle
- shle_acpp-ne
- shle_acpp
- shle_cpp-ne
- shle_cpp
- shle_ecpp-ne
- shle_ecpp
- shle_gpp
- x86
- x86_acpp-ne
- x86_acpp
- x86_cpp-ne
- x86_cpp
- x86_ecpp-ne
- x86_ecpp
- x86_gpp

The postfixes of the architecture are describing the runtime to be used. No postfix means the default what in turn is defined as "_cpp" ([see the -Y parameter](http://www.qnx.com/developers/docs/6.5.0/index.jsp?topic=%2Fcom.qnx.doc.neutrino_utilities%2Fq%2Fqcc.html)):
- _gpp — GNU C++ lib (available only for x86)
- _cpp — Dinkum C++ lib (default)
- _cpp-ne — Dinkum C++ lib (no exceptions)
- _acpp — Dinkum Abridged C++ lib
- _acpp-ne — Dinkum Abridged C++ lib (no exceptions)
- _ecpp — Embedded Dinkum C++ lib
- _ecpp-ne — Embedded Dinkum C++ lib (no exceptions)

QNX SDK 7.0
-------------
The SDK supports the following target-runtime-combinations out-of-the-box (printed by command: ```q++ -V```):
- aarch64le
- aarch64le_cxx
- aarch64le_gpp
- armv7le
- armv7le_cxx
- armv7le_gpp
- x86
- x86_64
- x86_64_cxx
- x86_64_gpp
- x86_cxx
- x86_gpp

The postfixes of the architecture are describing the runtime to be used. No postfix means the default what in turn is defined as "_cxx" ([see the -Y parameter](http://www.qnx.com/developers/docs/7.0.0/#com.qnx.doc.neutrino.utilities/topic/q/qcc.html)):
- _cxx — LLVM C++ library (default)
- _gpp — GNU C++ library

**Comment by SSE4** [comment]
I've checked, and `QCC` and `qcc` are identical binaries. it just check internally `argv[0]`, and adds appropriate language flags.

**Comment by Johnnyxy** [comment]
Request: change the default CMake flags to provide the CMake "Unix Makefile" generator instead of "MinGW Makefile" generator for QNX

code line to extend: [cmake_flags.py](https://github.com/conan-io/conan/blob/f1dc1cc74a54723220af9fa60ef4f0cd891fbb1c/conans/client/build/cmake_flags.py#L61)
```python
if os_build == "Windows" and not compiler == "qcc":
    return "MinGW Makefiles"  # it is valid only under Windows when not using QNX SDK compiler

return "Unix Makefiles"
```



Conan checks if the "os_build" is "Windows". If it is then Conan uses the "MinGW Makefiles".
But to compile with the QNX SDK compiler one needs the "Unix Makefiles" generator.
The reason behind this is that the QCC seems to be a "1:1" port from the GCC and does not support "MinGW Makefiles"-styled makefiles.

**Comment by SSE4** [comment]
@Johnnyxy or maybe even `compiler == "gcc"`? do "MinGW Makefiles" work with any other compiler except MinGW/GCC?

File: conans/client/migrations.py [hunk 4]
```diff
@@ -47,12 +47,12 @@ def _make_migrations(self, old_version):
             old_settings = """
 # Only for cross building, 'os_build/arch_build' is the system that runs Conan
 os_build: [Windows, WindowsStore, Linux, Macos, FreeBSD, SunOS]
-arch_build: [x86, x86_64, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr]
+arch_build: [x86, x86_64, ppc32be, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr, sh4le]
```
[the remark this answers is not in the document]
Comment (SSE4) [reply conans/client/migrations.py]: I am removing em

[the remark this answers is not in the document]
Comment (SSE4) [reply conans/client/migrations.py]: removed

File: conans/client/conf/__init__.py [hunk 5]
```diff
@@ -13,12 +13,12 @@
 default_settings_yml = """
 # Only for cross building, 'os_build/arch_build' is the system that runs Conan
 os_build: [Windows, WindowsStore, Linux, Macos, FreeBSD, SunOS]
-arch_build: [x86, x86_64, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr]
+arch_build: [x86, x86_64, ppc32be, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr, sh4le]
```
[the remark this answers is not in the document]
Comment (SSE4) [reply conans/client/conf/__init__.py]: what's the policy on adding them? we have bunch of watchOS/iOS related architectures here

**Comment by Johnnyxy** [comment]
Uh, I am no GCC guru :). I cannot say if this has to do something with GCC in general.
MinGW has a GCC Port too and uses its own Makefiles (which support Windows style commands).

I did not see any "pure" GCC compiler for Windows yet that uses ```Unix Makefiles```. Only those for MinGW ```MinGW Makefiles``` and QNX SDK ```Unix Makefiles```. 

Because of that I think the QNX guys "just" made the GCC compiler ready to be executed on Windows, but without any Windows specific additions thus using ```Unix Makefiles```.

For example: With the QNX SDK compiler I was not able to pass any argument with spaces to the compiler. Something like ```qcc c:/abc def/main.cpp```. The QCC will stop parsing after ```c:/abc``` and return an error that he cannot find the file ```c:/abc```.

If you enclose the path with quotes ```qcc "c:/abc def/main.cpp"``` the compiler returns an error that he does not understand the quote.

If you escape the path "linux-style" ```qcc c:/abc\ def/main.cpp``` the compiler returns an error that he cannot find the file "c:/abc\" (or something like that).

File: conans/client/conf/__init__.py [hunk 6]
```diff
@@ -13,12 +13,12 @@
...
 
 # Only for building cross compilation tools, 'os_target/arch_target' is the system for
 # which the tools generate code
-os_target: [Windows, Linux, Macos, Android, iOS, watchOS, tvOS, FreeBSD, SunOS, Arduino]
-arch_target: [x86, x86_64, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr]
+os_target: [Windows, Linux, Macos, Android, iOS, watchOS, tvOS, FreeBSD, SunOS, Arduino, Neutrino]
+arch_target: [x86, x86_64, ppc32be, ppc32, ppc64le, ppc64, armv6, armv7, armv7hf, armv7s, armv7k, armv8, armv8_32, armv8.3, sparc, sparcv9, mips, mips64, avr, sh4le]
```
Comment (zmostafa) [inline conans/client/conf/__init__.py]: Can you please add Little Endian arch for QNX targets , like : aarch64le , and armv7le?

Comment (SSE4) [reply conans/client/conf/__init__.py]: I think currently available ARM targets are already representing little endian ARM

Comment (Johnnyxy) [reply conans/client/conf/__init__.py]: In general the ARM architectures can use both BE and LE. And sometimes a mix.
The QCC compiler specifically state the the endianess for the ARM architectures.
e.g.:
QNX 7.0 and 6.5 -> armv7le
QNX 7.0 ->aarch64le

Comment (zmostafa) [reply conans/client/conf/__init__.py]: @Johnnyxy Thanks for the article,I think it would be nice to secify them in the arch_target at least for readability and complience with the qcc target names

Comment (Johnnyxy) [reply conans/client/conf/__init__.py]: No problem. I want QNX support too. :)

*_target settings
-----------------
@zmostafa you did not say for what you use the ```arch_target``` setting but be aware that you should not use the ```*_target``` settings in conan for normal compilation for you destination systems.
As the [documentation](https://docs.conan.io/en/latest/systems_cross_building/cross_building.html?highlight=gnu%20triplet#gnu-triplet-convention) states you should use the ```*_target``` settings only if you compile a software (os, arch) that compiles something for another target (os_target, arch_target) target.

If you compile a compiler on your build machine (Windows; x86_64) which runs on Linux (Linux; x86) that compiles for QNX (Neutrino; aarch64le) then your settings would look like this:
```python
os_build = "Windows"        # os that builds the compiler
arch_build = "x86_64"       # arch that builds the compiler
os = "Linux"                # os that runs the compiler
arch = "x86"                # arch that runs the compiler
os_target = "Neutrino"      # os that the built compiler should compile for
arch_target = "aarch64le"   # arch that the built compiler should compile for
```

If you compile only a piece of software (application, library) that runs on QNX (Neutrino, aarch64le) then you just use ```os``` and ```arch``` (maybe with os_build, arch_build _if nessescary_).
```python
os_build = "Windows"        # os that builds the application/library
arch_build = "x86_64"       # arch that builds the application/library
os = "Neutrino"             # os that runs the application/library
arch = "aarch64le"          # arch that runs the application/library
```

My recipes contains additionally to ```os``` and ```arch``` the ```*_build``` settings ```os_build```, ```arch_build``` because I have to distinguish between the QNX SDP for Windows and Linux (```os_build```).

Comment (zmostafa) [reply conans/client/conf/__init__.py]: @Johnnyxy Thanks for the clarification :) That means  "arch" line in the settings.yml has to be extended with QNX arch?

I am just building the applications, I have normal make files and Cmake and just want to use conan with those files.

Comment (Johnnyxy) [reply conans/client/conf/__init__.py]: Yes, exactly. If you just build libraries and applications then you have to set ```os = "Neutrino"```/```arch = "aarch64le"``` for your target system. I did this wrong in the beginning and had to do some refactoring because of my misunderstanding of the ```*_target``` settings.

> As general rule: ```*_target``` is almost never needed in the daily work of the most developers.

But some need this for special embedded boards or other rare systems or "simply" developing compilers so conan gives us this feature.

**Comment by Johnnyxy** [comment]
For reference: I share my boost/bzip2/zlib recipes, Conan profiles and CMake toolchains to make integration for QNX more transparent.
I have uploaded a ZIP file in this post: https://github.com/conan-community/community/issues/78#issuecomment-470857982

Additionally one can download the archive here directly.
[qnx_7_boost_bzip2_zlib_and_res.zip](https://github.com/conan-io/conan/files/2945127/qnx_7_boost_bzip2_zlib_and_res.zip)

**Comment by lasote** [comment]
I would like to push this forward, @SSE4 please, upgrade and resolve conflicts.
@zmostafa @Johnnyxy do you think this is ready to go?

**Comment by SSE4** [comment]
@lasote conflicts were resolved

File: conans/test/unittests/model/transitive_reqs_test.py [hunk 7]
```diff
@@ -1774,7 +1774,7 @@ def configure(self):
         self.assertEqual(node.ref, None)
         conanfile = node.conanfile
 
-        self.assertEqual(conanfile.version, "0.1")
+        self.equal = self.assertEqual(conanfile.version, "0.1")
```
Comment (memsharded) [inline conans/test/unittests/model/transitive_reqs_test.py]: weird

Comment (SSE4) [reply conans/test/unittests/model/transitive_reqs_test.py]: removed

**Comment by memsharded** [comment]
Hi all @zmostafa @Johnnyxy @jpovedano 

This has been merged, will be released in Conan 1.16. For any new issue regarding QNX support, please open new issues, thanks!
